exa/core/dao.py

```diff
@@ -0,0 +1,287 @@
+        dao.entities = ['id', 'user']
+        # if not specified, fetches all rows
+        dao.filters = {
+            'id': [('gt', 1), ('lt', 10)],
+            'user': [('eq', 'foo')]
+        }
+        df = dao(session=sqlalchemy.Session())
+
+    """
+    schema = Unicode(help='schema name')
+    table_name = Unicode(help='table name')
+    entities = List(help='columns to pull')
+    filters = Dict(help='filters to apply')
+
+    @default('entities')
+    def _default_entities(self):
+        return self.columns
+
+    @validate('schema', 'table_name')
+    def _validate_sql(self, prop):
+        sql = prop['value']
+        self.log.debug(f"validating {sql} against sql injection")
+        if any((i in sql for i in (';', '--'))):
+            raise SQLInjectionError("; or -- found. abort")
+        return sql
+
+    def _scan_sql(self, query):
+        self.log.debug("validating query against sql injection")
+        if any((i in query for i in (';', '--'))):
+            raise SQLInjectionError("; or -- found. abort")
+
+    def _filter_builder(self, query):
+        filters = []
+        for column, conditions in self.filters.items():
+            for op, cond in conditions:
+                fil = "{} {} '{}'".format(column, _raw_op_map[op], cond)
+                self.log.debug("building filter {}".format(fil))
+                filters.append(fil)
+        filters = ' and '.join(filters)
+        if filters:
+            query += ' where {}'.format(filters)
+        return query
+
+    def _query_builder(self, query):
+        entities = ', '.join(self.entities) or '*'
+        query = 'select {} from {}'.format(entities, self.fqtn())
+        query = self._filter_builder(query)
+        self._scan_sql(query)
+        return query + ';'
+
+    def _upload(self, session, payload, chunksize=_CHUNKSIZE):
+        if self.schema:
+            try:
+                session.execute(
+                    'create schema if not exists {};'.format(self.schema)
+                )
+                session.commit()
+            except Exception as e:
+                self.log.error("could not create schema: {}".format(e))
+        # TODO : implement and validate uniqueness constraints
+        #        on the Data class before insertion
+        nrec = len(payload.index)
+        if nrec > chunksize:
+            self.log.warning(f"loading {nrec} records at once")
+        bt = self.right_now()
+        payload.to_sql(
+            self.table_name, session.bind, index=False,
+            schema=self.schema, if_exists='append',
+        )
+        self.log.info(f"loading {nrec} took {self.time_diff(bt)}")
+
+
+    def __call__(self, session, payload=None, query_only=False):
+        if payload is not None:
+            return self._upload(session, payload)
+        query = self._query_builder('')
+        if query_only:
+            return query
+        self._data = pd.read_sql(query, session.bind)
```

Comment:
Why build raw sql and execute? read_sql function in pandas in pandas supports params argument

Reply:
the pandas API requires you to build the templated SQL along with the parameters you pass in. See https://stackoverflow.com/questions/24408557/pandas-read-sql-with-parameters for an example use case.

In the context of pd.read_sql, the functionality of the DAO is then to automate the process of building the SQL template strings, as well as composing params to conform to the underlying DB API spec. PEP 249 https://www.python.org/dev/peps/pep-0249/ outlines 5 distinct paramstyles that we would have to support for full compatibility.

The simpler option, for now, is to just build the raw SQL manually with no assumptions based on any dialect or API convention. I think it should certainly be enhanced in the future because the integration above will be more robust and feature complete. It just needs to be extensible for now so we can build this stuff in incrementally.

Reply:
Thanks for the thought out response. I'm okay with proposed approach.